### PR TITLE
Adds action to store media using spatie/laravel-medialibrary

### PR DESCRIPTION
**Pull request by core45**
**Purpose**

By default, the media files are stored in the `public` disk in the `media` directory. 

This adds the possibility to store media using `spatie/laravel-medialibrary` package.
The user will need to change the `media_action` key in the config file `filament-tiptap-editor.php` to `SpatieMediaLibraryAction`:

### CODE REVIEW

File: src/Actions/SpatieMediaLibraryAction.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                            if ($dimensions = $state->dimensions()) {
+                                $set('width', $dimensions[0]);
+                                $set('height', $dimensions[1]);
+                            }
+                        })
+                        ->saveUploadedFileUsing(function (BaseFileUpload $component, TemporaryUploadedFile $file, callable $set) {
+                            $model = $component->getContainer()->model;
+                            $media = $component->shouldPreserveFilenames()
+                                ? $model->addMedia($file)->toMediaCollection('tiptap-media')
+                                : $model->addMedia($file)->usingFileName(Str::uuid())->toMediaCollection('ck-media');
```
Comment (awcodes) [inline src/Actions/SpatieMediaLibraryAction.php]: What is 'ck-media'?

**Review by awcodes — COMMENTED** [review]
Overall, how are the fields like alt, etc getting saved or read from spatie's package? Since it doesn't support those as native fields on the record?

**Comment by core45** [comment]
Good points. Thank you.
I'll correct the ck-media thing and add alt etc

**Comment by core45** [comment]
I fixed the ck-media mistake.

Concerning the alt, title and dimensions there is no need to implement anything using spatie/laravel-medialibrary because those parameters are stored in the text itself as the  attributes eg. like this:

`Alt Text`

Is this OK?

**Comment by awcodes** [comment]
Wondering if this needs to be in the plugin itself? Seems like if could just be a Class in the app, since it can still be changed via the config.

**Comment by core45** [comment]
Well, you are the boss. But I think not. At lease no more than the "standard" way of storing files. I think adding this functionality do the core of the plugin will be beneficial taking under consideration how popular spatie/laravel-medialibrary is.
The very big downside to the standard way of storing files is that there is no control over the files (eg. converting/optimising  them) and even worse they are not deleted when the parent record is deleted.
I know it is possible to use your Curator plug-in but I can't do it because I already use spatie's one and changing this would be a nightmare. I believe this can be true for many many others.

**Comment by awcodes** [comment]
Valid points.

How would you handle inserting the converted images back into the editor? i.e, upload creates a 'medium' sized image, where would that option come to insert that size as the image in the editor. And also what if it takes a while to generate that image, you would then need a placeholder to show in the editor, which would require updates to the JS extension internally.

Not totally opposed to including this, I just think there's more to it that would require some further troubleshooting, or maybe even a better implementation of something like a 'Media Provider' class or similar.

**Comment by core45** [comment]
You are right - this can be a problem as everybody can name the conversion as they see fit...

The 'large' or 'medium' comes from the model.

For example:

```
 public function registerMediaConversions(Media $media = null): void
    {
        $this->addMediaConversion('thumb')->fit(Fit::Crop, 50, 50)->optimize()->format('webp');
        $this->addMediaConversion('preview')->fit(Fit::Crop, 450, 300)->optimize()->format('webp');
        $this->addMediaConversion('large')->fit(Fit::Crop, 850, 548)->optimize()->sharpen(10)->format('webp');
    }

public function getFeaturedImageAttribute()
    {
        $file = $this->getMedia('featured_image')->last();

        if ($file) {
            $file->url       = $file->getUrl();
            $file->thumbnail = $file->getUrl('thumb');
            $file->preview   = $file->getUrl('preview');
            $file->large     = $file->getUrl('large');
        }

        return $file;
    }
```

So using my spatie solution you should be able to somehow decide which conversion (if any) you want to use.

I think this should be set by adding  something like this (so with every model could use different conversion):

TiptapEditor::make('main_text')->useSpatieConversion('large'),

Or add it in setting file to use in all models.

What do you think?

**Comment by awcodes** [comment]
Honestly, I think this is assuming a lot. But I do see the value, despite knowing, at least right now, the best way forward. Like I'm torn on supporting something I don't personally use, but at the same time it's kind of the point of being able to provide your own custom Media class.

**Comment by awcodes** [comment]
I'm going to close this for now. Thank you for your hard work on this. I will consider it more for v4, but I feel that this is still better handled outside the core plug-in.

If you'd like to release it as an add on for Curator, I'd be OK with it. Cheers.

**Pull request closed without merging**